### PR TITLE
Fix object controls events

### PR DESCRIPTION
**Pull request by Hadegda**
## Description
Review objectControls events and fix problems with them

## Type of changes
- New feature (non-breaking change which adds functionality)

## Checklist
- [x] I have read [CONTRIBUTING](https://github.com/epam/miew/blob/master/CONTRIBUTING.md) and [CODE_OF_CONDUCT](https://github.com/epam/miew/blob/master/CODE_OF_CONDUCT.md) guides.
- [x] I have followed the code style of this project.
- [x] I have run `yarn run ci`: lint and tests pass locally with my changes.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary documentation / The changes do not need docs update.

### CODE REVIEW

File: src/Miew.js [hunk 1]
```diff
@@ -2930,8 +2930,9 @@ Miew.prototype.resetPivot = function () {
     boundingBox.union(visual.getBoundaries().boundingBox);
   });
 
-  boundingBox.getCenter(this._gfx.pivot.position);
-  this._gfx.pivot.position.negate();
+  const center = new THREE.Vector3();
```
Comment (NataliaDSmirnova) [inline src/Miew.js]: Hi, I  have a question about new object creation. How often is the function called? Could we avoid using of `new` every time?

File: src/Miew.js [hunk 2]
```diff
@@ -2941,9 +2942,9 @@ Miew.prototype.setPivotResidue = function (residue) {
     return;
   }
 
-  const pos = this._gfx.pivot.position;
+  const center = new THREE.Vector3();
```
Comment (NataliaDSmirnova) [inline src/Miew.js]: Here is the same point about vector creation. But I agree the modification of the temporary variable is safer than a change of pivot's component.

File: src/Miew.js [hunk 3]
```diff
@@ -2968,10 +2969,9 @@ Miew.prototype.setPivotAtom = function (atom) {
     return;
   }
 
-  const pos = this._gfx.pivot.position;
-  pos.copy(atom._position);
-  pos.applyMatrix4(visual.matrix);
-  pos.negate();
+  const center = atom._position.clone();
```
Comment (NataliaDSmirnova) [inline src/Miew.js]: One again, clone() means newly creation, isn't it?

File: src/ui/ObjectControls.js [hunk 4]
```diff
@@ -68,12 +68,14 @@ ObjectHandler.prototype.setObjects = function (objects) {
 
 ObjectHandler.prototype.rotate = function (mousePrevPos, mouseCurPos, aboutAxis) {
   const rot = this.mouse2rotation(mousePrevPos, mouseCurPos, aboutAxis);
+  const quat = new THREE.Quaternion().setFromAxisAngle(rot.axis, rot.angle);
```
Comment (NataliaDSmirnova) [inline src/ui/ObjectControls.js]: I see you've left a quaternion creation that has already existed. But can we refactor the function using closure? What do you think?

File: src/ui/ObjectControls.js [hunk 5]
```diff
@@ -543,8 +547,9 @@ ObjectControls.prototype.update = function () {
           this._altObj.axis.set(0, 0, 1);
         }
 
-        this._altObj.translate(new THREE.Vector2(delta * speedX, delta * speedY));
-        this.dispatchEvent({ type: 'change', action: 'translate' });
+        const shift = new THREE.Vector2(delta * speedX, delta * speedY);
```
Comment (NataliaDSmirnova) [inline src/ui/ObjectControls.js]: Here again 😄

File: src/ui/ObjectControls.js [hunk 6]
```diff
@@ -785,14 +787,16 @@ ObjectControls.prototype.translatePivotByMouse = function () {
 // Translate in WorldCS, translation is scaled with root scale matrix
 ObjectControls.prototype.translatePivotInWorld = function (x, y, z) {
   const pos = this.objectPivot.position;
+  const oldPos = this.objectPivot.position.clone();
```
Comment (NataliaDSmirnova) [inline src/ui/ObjectControls.js]: and here 😸

File: src/ui/ObjectControls.js [hunk 7]
```diff
@@ -785,14 +787,16 @@ ObjectControls.prototype.translatePivotByMouse = function () {
...
   pos.applyMatrix4(this.object.matrixWorld);
   pos.setX(pos.x + x);
   pos.setY(pos.y + y);
   pos.setZ(pos.z + z);
   const invWorldMat = new THREE.Matrix4().getInverse(this.object.matrixWorld);
```
Comment (NataliaDSmirnova) [inline src/ui/ObjectControls.js]: Hmmm, and here.

File: src/ui/ObjectControls.js [hunk 8]
```diff
@@ -785,14 +787,16 @@ ObjectControls.prototype.translatePivotByMouse = function () {
...
   pos.applyMatrix4(invWorldMat);
 
-  this.dispatchEvent({ type: 'change' });
+  const shift = oldPos - pos;
```
Comment (NataliaDSmirnova) [inline src/ui/ObjectControls.js]: Does difference operator exist for the Vector3?

File: src/Miew.js (hunk 1 above)
Comment (Hadegda) [reply src/Miew.js]: Thanks for your comment, done

File: src/Miew.js (hunk 2 above)
Comment (Hadegda) [reply src/Miew.js]: done

File: src/Miew.js (hunk 3 above)
Comment (Hadegda) [reply src/Miew.js]: Yes, change it

File: src/ui/ObjectControls.js (hunk 4 above)
Comment (Hadegda) [reply src/ui/ObjectControls.js]: I think it's great idea
Done)

File: src/ui/ObjectControls.js (hunk 5 above)
Comment (Hadegda) [reply src/ui/ObjectControls.js]: done)

File: src/ui/ObjectControls.js (hunk 7 above)
Comment (Hadegda) [reply src/ui/ObjectControls.js]: It looks like the previous note)

File: src/ui/ObjectControls.js (hunk 6 above)
Comment (Hadegda) [reply src/ui/ObjectControls.js]: done)

File: src/ui/ObjectControls.js (hunk 8 above)
Comment (Hadegda) [reply src/ui/ObjectControls.js]: Yes, thank you
I fix it

**Comment by NataliaDSmirnova** [comment]
Thank you for the job 👍

**Pull request merged**